### PR TITLE
Refactor to encapsulate strain logic into Skill class

### PR DESCRIPTION
**Pull request by Syriiin**
This PR is part of #11407 which I have split into 4 smaller PRs for the sake of backporting to osu-stable in smaller chunks.
- Add a list of mods to `Skill` class (#11687)
- Implement dynamic previous hitobject retention for `Skill` class (#11688)
- Refactor to encapsulate strain logic into `Skill` class (#11689)
    - Refactor to abstract out strain logic into `StrainSkill` class (#11690, branched from #11689)

# Changes

- Refactors to encapsulate the strain logic into the `Skill` class rather than at the `DifficultyCalculator` level

# Reason

As strains are an implementation detail of the current skill calculations, it makes sense that strain related logic should be encapsulated within the `Skill` class.

Depends on #11849

### CODE REVIEW

File: osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs
Comment (bdach) [inline osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs:36]: This change (and all the other similar changes in this file) are not actually equivalent. The previous value (read from `BaseObject`) was independent of track rate. This one is not.

File: osu.Game.Rulesets.Taiko.Tests/TaikoDifficultyCalculatorTest.cs
Comment (bdach) [inline osu.Game.Rulesets.Taiko.Tests/TaikoDifficultyCalculatorTest.cs:17]: Test adjustments like that in refactoring PRs should come with a proper rationale (which I've provided for you in another review comment).

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/Rulesets/Difficulty/Skills/Skill.cs
Comment (smoogipoo) [reply osu.Game/Rulesets/Difficulty/Skills/Skill.cs:119]: I'm not sure I understand this comment. The final strain peak is already added at the end:
https://github.com/ppy/osu/blob/e8f08b09dcc302652a66b5476fe9c1fd7ec8e43a/osu.Game/Rulesets/Difficulty/DifficultyCalculator.cs#L96-L98

If I'm missing something here then this is a huge bug, because I've already had to address this in osu-stable too to match lazer (with a completely different diffcalc structure).

Comment (bdach) [reply osu.Game/Rulesets/Difficulty/Skills/Skill.cs:119]: Yes, and current `master` is correct. What I'm getting at is that taiko diffcalc at one point queries the skills' `StrainPeak`s, and due to this code not saving the last peak to that protected list as was done previously (to "avoid side effects"), it changes taiko results.

Comment (Syriiin) [reply osu.Game/Rulesets/Difficulty/Skills/Skill.cs:124]: Ah I see, nice catch.
Fixed in 3743937774fb21887ee2472555274a5784d44048

I've opted to expose `CurrentStrainPeak` as a property as opposed to modifying `StrainPeaks` to add the final incomplete section's peak for 2 reason.

1. A dynamic `StrainPeaks` property could be a little confusing if other expect it to directly reflect the internal `strainPeaks` field.
2. It seems like a bad idea to copy an entire list for the sake of a dynamic property when efficiency is a key objective.

File: osu.Game.Rulesets.Taiko.Tests/TaikoDifficultyCalculatorTest.cs
Comment (Syriiin) [reply osu.Game.Rulesets.Taiko.Tests/TaikoDifficultyCalculatorTest.cs]: Fixed in 3743937774fb21887ee2472555274a5784d44048
see https://github.com/ppy/osu/pull/11689#discussion_r571817264

File: osu.Game/Rulesets/Difficulty/Skills/Skill.cs
Comment (bdach) [reply osu.Game/Rulesets/Difficulty/Skills/Skill.cs:124]: You could use `Enumerable.Append` to avoid a copy.

Not super sure it's reasonable to make consumers remember to include the last peak themselves, to be honest.

File: osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs
Comment (Syriiin) [reply osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs]: Ah yea that's totally right. I wonder why I did this in the first place with the original PR 🤔 
It's been a while since I made that change, but I probably thought it was a bug since its counter intuitive that mania calculations would just ignore clockrate like that.

Fixed in ff08eee43b71eeaeb1e16d2d5d8c7f9e321e8899

Out of curiosity, do you know why the mania calc does this?
The whole idea behind clockrate being passed to the DifficultyHitObject abstraction is to remove differences between a clockrate adjusted map and the same objects timing existing as nomod.

File: osu.Game/Rulesets/Difficulty/Skills/Skill.cs
Comment (Syriiin) [reply osu.Game/Rulesets/Difficulty/Skills/Skill.cs:119]: Ooh never knew about that method. Pretty neat.
Yea, I'm happy to use that instead. I'll change the name of the property though to make it more obvious that its more nuanced than a mere getter.

Fixed in 01bde91a5b2b1c9aea51c434fe1426042e2b8de9

File: osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs
Comment (bdach) [inline osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs:152]: This code is too ugly to live. I did not anticipate that changing the peaks to `IEnumerable` would cause this, but in light of it I'd rather have a few redundant list copies in order to be able to access by index normally.

Either that, or do something with `.Zip()`, but I think any solution that would use that is likely to be too ugly as well.

File: osu.Game/Rulesets/Difficulty/Skills/Skill.cs
Comment (bdach) [inline osu.Game/Rulesets/Difficulty/Skills/Skill.cs:119]: This can (and should) just use `CurrentStrainPeaks` now.

**Review by bdach — DISMISSED** [review]

File: osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs
Comment (bdach) [reply osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs]: `ManiaDifficultyCalculator` itself includes adjustments for rate-changing mods.

File: osu.Game/Rulesets/Difficulty/Skills/Skill.cs
Comment (Syriiin) [reply osu.Game/Rulesets/Difficulty/Skills/Skill.cs]: Fixed in 2f8fd1ceea8d5787d04985b9a70e724ae76a45a4

File: osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs
Comment (Syriiin) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs]: I agree it's not as elegant as indexed for loop, or foreach with `Zip()` but I certainly don't agree that it's "too ugly to live'...
By all rights, this is a textbook intended usage of multiple low level enumerators.
I've cleaned it up slightly in 05278e0510ef2b0492c41296978cb585811e7901 by using `using` blocks to avoid the `Dispose()` calls.
I'd be keen to get other's thought on this, because I personally don't think this is a good enough reason to justify intentional efficiency impacts (no matter how small) on an efficiency sensitive part of the codebase.

Comment (bdach) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs]: If performance is a worry then the only thing that will dispel any and all doubts is a benchmark, and not theory crafting.

Comment (Syriiin) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs]: My point is that the reason we started down this path in the first place was because you didn't want the needless copy https://github.com/ppy/osu/pull/11689#discussion_r571820236.
I'm still of the opinion that the best method is either what I originally did https://github.com/ppy/osu/pull/11689#discussion_r571817264, or the enumerable approach, since a copy in the property is an unexpected allocation. There's no need to benchmark, since the difference is literally sub-millisecond, but its about the principle.
Regardless, I'm tired of arguing implementation details. I've changed it to have a copy in the property.

Changed in b6ac32bddd5bc05a215b3710ea1dbfcf44ba1a89

File: osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs
Comment (Syriiin) [reply osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs]: That feels like something that should be addressed.
It very much feels like the difficulty calculator framework was built for osu standard and instead of modifying it to better support the needs of other rulesets, instead other ruleset calculators have just been shoehorned into it and made hacky solutions to fix issues (another example being the taiko `Rhythm` skill's index based strains).
Those are not issues for this PR though.

Comment (smoogipoo) [reply osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs]: The simple reason why mania does this is that it was build this way from the very beginning. It's not meant to be bug-less, it's meant to replace osu-stable first and foremost to provide a path forward for which such bugs can be fixed and meaningful comparisons can be made against osu-stable.

Comment (Syriiin) [reply osu.Game.Rulesets.Mania/Difficulty/Skills/Strain.cs]: Ah yea that makes sense.
For some reason I thought there had been a major mania restructure after it was ported to lazer. 🤔

File: osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs
Comment (bdach) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs]: For the record: The primary reason I requested this change is that using raw enumerators is rather rare, using four at a time like the previous code did - especially so. I think making some list copies is a small price to pay in exchange for adhering to the principle of least astonishment. Feel free to overrule me on this, but I don't think the impact of the list copies is large enough to justify the code being non-obvious to comprehend.

**Comment by bdach** [comment]
I have made several follow-up changes that bring this code more in line with how I'd expect it to look. Explanations for the changes (if necessary) are in the commit messages.

I have no more objections with this code anymore, so dismissing my review, but definitely a second look would be pertinent.

File: osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs
Comment (Syriiin) [reply osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs]: > using raw enumerators is rather rare, using four at a time like the previous code did - especially so

Rare circumstances require rare techniques.

While I do disagree with your opinion here, given that you have already gone ahead and committed further changes on top of this by yourself, I don't think its worth the effort to push the issue any further.

I'm happy to leave it as you have committed, however I do stand by my opinion that what I originally did in 3743937774fb21887ee2472555274a5784d44048 is the most elegant solution; also adhering to the principle of least astonishment.

**Comment by Syriiin** [comment]
> 19485429dc70b07ad333d27ebf3f88f2b4530f48
Also implements the property in a less opaque manner (previous
implementation required the reader to be aware that object initialisers
are syntactic sugar for `.Add()` calls, which does not readable code
make).

As I mentioned in https://github.com/ppy/osu/pull/11689#discussion_r576667235, I'm happy to leave these changes as is, but I wanted to point out that I think its rather strange to suggest elegant syntactic sugar - like object initialisers - shouldn't be used because the reader may not be aware it exists.
I would understand if this was some cryptic oneliner, but I would think this usage was fairly obvious given the variable naming and clear explanation in the xmldoc.

> cfa3db3b78276307db27484333d8f9ccde3fcf3a
Inadvertently changed from the rate-independent value of the start time
to the rate-dependent one.

This was not an inadvertent change.
This reversion completely breaks clock rate adjusted difficulty calculation.
It is a required change to account for the fact that we no longer are scaling the `SectionLength` with the clock rate, since `Skill`s should be unaware of such things in the first place.

Fixed in 05854b83d83b9f7ce85bf29ea99d9e5ef8f6b47b

**Comment by bdach** [comment]
> I think its rather strange to suggest elegant syntactic sugar - like object initialisers - shouldn't be used because the reader may not be aware it exists.

I'm fine with object initialisers for empty collections. I'm not fine with using them alongside copy constructors.

**Comment by Syriiin** [comment]
I was just working on adding test cases for clock rate adjusted difficulty calculation (since they would have helped avoid the earlier accidental reversion of the required `GetPeakStrain()` change), and noticed that clock rate adjustment calculations are completely broken for mania.
I believe this is actually the reason I had made the change referred to here https://github.com/ppy/osu/pull/11689#discussion_r571823708 and had forgotten about. 😅

With that change added back in, it looked like the values came back very close to master (0.02 stars difference from the one test case I had iirc).

My house just had a blackout though so I couldn't finish investigating.
Posting this incase someone wants to look into this before my power comes back, whenever that might be...

**Comment by Syriiin** [comment]
re: https://github.com/ppy/osu/pull/11689#issuecomment-779795402

I have investigated this further and I believe there is actually a bug with the existing mania calculator on master.

Running the calculator against mania's `diffcalc-test.osu` with DT enabled reveals a 0.02 star error compared to a base clockrate calculation of [the same map with its timings manually adjusted for DT](https://gist.github.com/Syriiin/c14090a6d38fd9e7b9c9db1b4cd10d6f).
Running the same test on this branch, with the fixes referred to in https://github.com/ppy/osu/pull/11689#discussion_r571823708 that I have added back in with f1d964734e6d19aa70a17d2fb584875d2810efb5, yields only an expected 0.0001 star rounding error.

```
master:
      Expected: 2.7880168137446191d
      But was:  2.7646128945056723d
refactor:
      Expected: 2.7880168137446191d
      But was:  2.7879104989252959d

master:   0.023403919238946802 error
refactor: 0.00010631481932321662 error
```

This is of course all predicated on this actually being a bug and that the intention is for mania maps with DT to be equivalent in difficulty calculations to the same map recreated in nomod. I don't see any reason why this _would_ be the case, since it makes no logical sense to me, but maybe I'm missing something.

@smoogipoo would you be able to confirm whether this is a real bug and solution?

**Comment by bdach** [comment]
I'll say this again: *this pull should not be changing **any** diffcalc values*. It's supposed to be a code-level restructuring. Please stop with the scope creep. Was that not the reason to split this in the first place?

It may well be that there is a "bug" in mania. But if there is, it needs to be checked if the same bug is in stable. And if it also is present in stable, then it ought to be fixed in both. Otherwise that bug being replicated is a *feature* until the issue is fixed in both places.

**Comment by Syriiin** [comment]
> I'll say this again: _this pull should not be changing **any** diffcalc values_. It's supposed to be a code-level restructuring. Please stop with the scope creep.

Please read what I write more carefully before making comments like this.
If you had, you would know that there is no way to implement these changes without fixing this bug.
The strain refactor completely breaks the manic calc, so this is a REQUIRED change.
Personally I find the pedantry this sort of comment entails to be over the top and counter productive, but what do I know? That's just my 2c.

> It may well be that there is a "bug" in mania. But if there is, it needs to be checked if the same bug is in stable. 

Yes I bug does exist in stable.
Loading the same map files from https://github.com/ppy/osu/pull/11689#issuecomment-780662699 into stable will reveal the same error margin, as well as a possibly unrelated difference compared to master lazer values for the same maps.

`diffcalc-test-dt.osu` (manual timing adjustments)
image
image

`diffcalc-test.osu` +DT
image
image

> And if it also is present in stable, then it ought to be fixed in both separately. That bug being replicated is a feature until the issue is fixed in both places.

Okay. I will create another PR to specifically address this bug and rebase this branch onto that one.

**Comment by bdach** [comment]
> If you had, you would know that there is no way to implement these changes without fixing this bug.

As far as I can see this conclusion does not follow from anything that I've read in this thread. It was already stated previously that diffcalc code is supposed to primarily match stable for the time being.

**Comment by smoogipoo** [comment]
If there is absolutely no way to make forward progress without breaking stable, then so be it. It's quite bad being in a "refactor" PR which isn't supposed to change logic, though.

But just keep in mind that this is going to delay this PR from something that could be merged relatively quickly to now an indeterminate amount of time, because we need to make sure that not only does stable reflect these changes but also that lazer isn't broken across the entire beatmap library and all modes along with it _in comparison to stable_. I also need to confirm that the changes are non-regressing and the changes are expected, rather than doing a simple lazer-master vs lazer-pr comparison across the beatmap library and making sure there's no changes.

Unfortunately this has to be done because diffcalc is our best bet for finding and fixing issues where lazer differs in, say, beatmap conversion vs stable.

I currently don't have the capacity to do all of this and upkeep the development of the game, but as long as you understand why it'll delay this PR then no problem.

**Comment by Syriiin** [comment]
> As far as I can see this conclusion does not follow from anything that I've read in this thread.

https://github.com/ppy/osu/pull/11689#issuecomment-779795402 + https://github.com/ppy/osu/pull/11689#issuecomment-780662699
Perhaps I should have mentioned this more explicitly.
Nevertheless, this is an unproductive tangent.

> https://github.com/ppy/osu/pull/11689#issuecomment-781238199

Yep, all good.
I'll do what I mentioned and try to split this fix out into it's own PR to rebase this branch onto.

**Comment by Syriiin** [comment]
I investigated further on this bug and PR'd a fix for it in #11849
It turns out it was actually a bug affecting all gamemodes.
I have rebased this branch onto that PR's branch now and squashed the changes of this branch to clean up the commits.

**Comment by bdach** [comment]
If it affected all gamemodes, then if I understand correctly, then the rate fix is entirely orthogonal to this pull. By which I mean, this refactoring *can* go through without fixing that issue, as long as the fix is reverted here.

And yes, it may well be wrong. Again, I'm not interested in discussing that in the scope of this pull. It's just that if my above understanding is correct, I do not see the reason to hold this PR up longer than it needs to.

**Comment by Syriiin** [comment]
> If it affected all gamemodes, then if I understand correctly, the rate fix is entirely orthogonal to this pull. By which I mean, this refactoring _can_ go through without fixing that issue, as long as the fix is reverted here (which I attempted to do, but without being aware of the issue).

Unfortunately not quite, I don't think.
The reason I noticed this issue in the first place was because this refactor fixed the issue by its nature, since it moved the bugged logic to the `Process()` call where the timings were now clockrate adjusted.
Unless I'm overlooking something, I don't believe there is any way for this refactor to complete its goal of moving the strain logic into the Skill class, without consequently fixing the decay bug and changing diffcalc values.

**Comment by Syriiin** [comment]
In the meantime though, there's still the independent pure refactors in #11687 and #11688 which were reviewed as part of the original PR and should be just about ready for approval.

**Comment by bdach** [comment]
There's another pull that makes this same sort of change, and it does this: https://github.com/Xexxar/osu/blob/a154cdfb7914daa4fa0c3889a5814f2c0e2eb680/osu.Game/Rulesets/Difficulty/Skills/Skill.cs#L64

Is that not a viable way forward? Is this what you were talking about when you said

> It is a required change to account for the fact that we no longer are scaling the `SectionLength` with the clock rate, since `Skill`s should be unaware of such things in the first place.

Why should skills not be aware of the clock rate? I don't see a huge reason for them to not be aware of it.

**Comment by Syriiin** [comment]
> There's another pull that makes this same sort of change, and it does this: https://github.com/Xexxar/osu/blob/a154cdfb7914daa4fa0c3889a5814f2c0e2eb680/osu.Game/Rulesets/Difficulty/Skills/Skill.cs#L64
> 
> Is that not a viable way forward? Is this what you were talking about when you said
> 
> Why should skills not be aware of the clock rate? I don't see a huge reason for them to not be aware of it.

Ah yes, I knew it would be *possible* to do it this way, by passing the clockrate on each `DifficultyHitObject`, or directly to `Process()`, but I really *really* think its an antipattern that we ought to avoid. In fact, I have another PR open now to refactor this sort of clockrate usage out of the catch calculator.

The original and intended purpose of the `DifficultyHitObject` abstraction was (and still should be) to represent the playable objects of a beatmap in a given context, *without exposing what that context is* so they can be treated uniformly independent of what transformations (as a result of mods, etc...) brought them to this state.
If this pattern was adhered to 100% (it is, except for clockrate and base timing usages right now afaik), then the difficulty calculators would be guaranteed to be consistent for equivalent playable beatmaps; making the sort of bugs I've been encountering recently impossible.

Adding clockrate to `DifficultyHitObject`s is a step backwards and defeats the purpose of this abstraction.
If we really just want to rush this PR ahead, I would be okay with doing this temporarily, but I would be making another PR to refactor to bring the clockrate back to the `DifficultyCalculator` level if that were the case.
I think this isn't a great idea, and we should focus on fixing the bug first, but this option exists if we really want it.

**Comment by bdach** [comment]
As far as I can tell there's nothing stopping someone from accessing `DifficultyHitObject.BaseObject`, which would have the rate-unadjusted start time and duration. So I don't know that those sorts of bugs would be impossible - not without hiding that information somehow, that is.

**Comment by Syriiin** [comment]
> As far as I can tell there's nothing stopping someone from accessing `DifficultyHitObject.BaseObject`, which would have the rate-unadjusted start time and duration. So I don't know that those sorts of bugs would be impossible - not without hiding that information somehow, that is.

Well yes that's exactly right. Ideally we shouldn't be exposing `BaseObject` at all.
Once these changes are merged, there will be very little, if anything, stopping us from removing `BaseObject` entirely.
I plan to make a PR for this after this one has been ironed out.

**Comment by smoogipoo** [comment]
Need a bit of conflict resolution here.

**Comment by Syriiin** [comment]
> Need a bit of conflict resolution here.

👍

**Comment by bdach** [comment]
As I said on discord, resolving merge conflicts is not the same thing as force-pushing a squashed version of the entire branch. Please refrain from performing destructive operations on git history unless absolutely necessary, as the [contributing guidelines](https://github.com/ppy/osu/blob/master/CONTRIBUTING.md) already state quite clearly.

**Comment by smoogipoo** [comment]
Sheet of top-10000 changes: https://docs.google.com/spreadsheets/d/1dHzGN7PBkh2RLc6jkdOS0Ea-C8o5M3ewfHeivbyXJ2s

Concentrated around mania, due to the clock-adjusted endtime/starttime changes. But the changes look reasonable to me.

**Review by smoogipoo — APPROVED** [review]
Lgtm

**Pull request merged**
